indra/assemblers/html/assembler.py

```diff
@@ -174,163 +202,222 @@ def make_json_model(self, with_grouping=True, no_redundancy=False):
             various metadata.
         """
         # Get an iterator over the statements, carefully grouped.
-        stmt_rows = group_and_sort_statements(
-            self.statements,
-            self.ev_counts if self.ev_counts else None,
-            self.source_counts if self.source_counts else None)
-
-        # Do some extra formatting.
-        stmts = OrderedDict()
+        normal_stats = make_standard_stats(ev_counts=self.ev_counts,
+                                           beliefs=self.beliefs,
+                                           source_counts=self.source_counts)
+        stats = normal_stats + self.custom_stats
+        stmt_rows = group_and_sort_statements(self.statements,
+                                              custom_stats=stats,
+                                              sort_by=self.sort_by,
+                                              grouping_level=grouping_level)
+
+        # Set up some data structures to gather results.
         agents = {}
-        previous_stmt_set = set()
-        all_previous_stmts = set()
-        for row in stmt_rows:
-            # Distinguish between the cases with source counts and without.
-            if self.source_counts:
-                key, verb, stmts_group, tl_counts, src_counts = row
-            else:
-                key, verb, stmts_group = row
-                src_counts = None
-                tl_counts = None
-            curr_stmt_set = {s.get_hash() for s in stmts_group}
-            if curr_stmt_set == previous_stmt_set:
-                continue
-            elif no_redundancy and curr_stmt_set <= all_previous_stmts:
-                continue
+        source_count_keys = set() if not self.source_counts \
+            else {k for k in next(iter(self.source_counts.values())).keys()}
+
+        # Loop through the sorted and grouped statements.
+        all_hashes = set()
+
+        def handle_rows(rows, level, **kwargs):
```

Comment:
Instead of kwargs, probably better to just add the explicit optional arguments `agent_key` and `agp_agents`, because those are the only ones that are ever provided to the function.

Reply:
They aren't the only arguments; they change depending on which level you are calling. I think putting in the rather arbitrary other arguments would actually serve to obfuscate the meaning of the function. `handle_rows` takes arguments of "rows", "level", and "maybe some other things". I really think that is the best way to think of this function.